### ISSUE TITLE
Publish cosmos libraries to nuget.org

### ISSUE THREAD

**Issue by terminal-cs** [CosmosOS/Cosmos#2608]
#### Area of Cosmos - What area of Cosmos are we dealing with?
Cosmos/General

#### Expected Behaviour - What do you think that should happen?
n/a

#### Actual Behaviour - What unexpectedly happens?
n/a

#### Reproduction - How did you get this error to appear?
n/a

#### Version - Were you using the User Kit or Dev Kit? And what User Kit version or Dev Kit commit (Cosmos, IL2CPU, X#)?
n/a

#### Reason/Explanation

We should be using nuget.org to publish cosmos packages. Ideally, we should have a GitHub action run every time a PR is merged to update the libraries on nuget.org. This would both allow users to get instant updates to the core libraries without having to update cosmos every time, And it would also greatly ease using cosmos projects in testing/documentation software via GitHub actions, so that cosmos doesn't need to be set up again every time an action runs. Optionally we could have 2 channels (stable & developer), however, this may be a bit more complicated. If it is done, we would need to make sure that existing projects switch to and from the stable and dev channels correctly when loaded when either the user-kit or the dev-kit is installed.

Edit: This would also improve the usability of cosmos on linux because we won't have to worry about local-only nuget packages being installed properly, which has issues currently.

**Comment by zarlo** [comment]
we will not likely do this any time soon as we use some hard coded paths and it would not work as cosmos still needs to be installed as not every part of it can be made into a nuget package

**Comment by quajak** [comment]
Is it possible to include everything which is currently stored in the User kit directory in a nuget package?

**Comment by charlesbetros** [comment]
We will most likely start with publishing packages to our feed in GitHub rather than nuget.org. The kernel libraries are already nuget packages and are installed to a local nuget feed. They are hood candidates to move out of the user kit and in to our GitHub feed.

**Comment by terminal-cs** [comment]
> Is it possible to include everything which is currently stored in the User kit directory in a nuget package?

All of the DLLs yes

**Comment by quajak** [comment]
Yeah thinking about it more, there is no real advantage to the local package source. If i remember correctly, the GitHub nuget package source is very difficult to use since you cannot make it publicly available (you need access rights) but they might have finally changed that.

**Comment by zarlo** [comment]
bit i think we should start building X# as a nuget package to make it easier to use i got this to work in on linux but could not get it work with the windows installer builder the patch file https://gitlab.com/zarlo/cosmos.zarlo/-/blob/main/Patches/Cosmos/makefile-xsharp.patch

**Comment by terminal-cs** [comment]
> > Is it possible to include everything which is currently stored in the User kit directory in a nuget package?
> 
> depends on if we can call a process from inside a nuget package (calling IL2CPU)

Not the process, the libraries

**Comment by terminal-cs** [comment]
@quajak we should bundle il2cpu, both libraries and the executable in its own nuget package. It's possible to store and run EXEs in nuget packages

**Comment by terminal-cs** [comment]
![Screenshot_20230316-162149~2](https://user-images.githubusercontent.com/76945439/225773866-19813c23-b37d-4949-be1e-3921de9a5509.png)

More info that I SSed, just some ideas

**Comment by zarlo** [comment]
okay another thing we need to do it load plugs from nuget packages not from the install folder

**Comment by zarlo** [comment]
> Yeah thinking about it more, there is no real advantage to the local package source. If i remember correctly, the GitHub nuget package source is very difficult to use since you cannot make it publicly available (you need access rights) but they might have finally changed that.

this is mostly just a note:

but i have a way round this that i use on gitlab that would work on github

i made a cicd pipeline that runs every 6 hours and it uses sleet to make a static nuget repo and it have it pull in from the gitlab nuget repos

https://gitlab.com/zarlo/nuget

**Comment by valentinbreiz** [comment]
Cosmos Gen3 has replaced the IL2CPU-based Gen2 codebase, and `gen3` is now the default branch of this repository. This issue was filed against Gen2, which is no longer developed, so it is being closed as part of the Gen3 cutover (#3088).

If you still hit this on Gen3, please open a fresh issue. The compiler, the build pipeline and the debugging story are all different now, so a new report will get a far more useful answer than reviving this one. Start here: https://cosmosos.github.io/

Gen2 remains on the `gen2` branch, and this issue stays readable.
